README.md
# Ystia Forge

[![Download](https://img.shields.io/badge/download-v3.0.0--milestone.1-blue)](https://github.com/ystia/forge/releases/tag/v3.0.0-milestone.1) [![Build Status](https://github.com/ystia/forge/actions/workflows/build.yml/badge.svg?branch=develop)](https://github.com/ystia/forge/actions) [![license](https://img.shields.io/github/license/ystia/forge.svg)](https://github.com/ystia/forge/blob/develop/LICENSE) [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com) [![Join the chat at https://gitter.im/ystia/forge](https://badges.gitter.im/ystia/forge.svg)](https://gitter.im/ystia/forge?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

A set of public TOSCA Types and Topology templates used by the Ystia project

## Documentation

See [our documentation for more details](https://github.com/ystia/forge/blob/develop/org/ystia/README.rst)

## Downloads

While Alien4Cloud has a support for importing types directly from Git, you can still visit our [GitHub releases page](https://github.com/ystia/forge/releases) for downloading a release.

Grab the [latest release here](https://github.com/ystia/forge/releases/latest).


CHANGELOG.md
# Ystia Forge Changelog

## UNRELEASED

### ENHANCEMENTS

* Add missing hostname in yorc certificate Subject Alternative Name ([GH-151](https://github.com/ystia/forge/issues/151))

### BUG FIXES

* Docker installation requires apt-transport-https to work properly on Debian based systems ([GH-147](https://github.com/ystia/forge/issues/147))
* [Yorc] Error in AWS location configuration ([GH-145](https://github.com/ystia/forge/issues/145)) | Thanks to [@tibeer](https://github.com/tibeer) for this contribution

## 3.0.0-milestone.1 (April 25, 2021)

### DEPENDENCIES

* Ystia Forge components require now Alien4Cloud 3.0.0

### ENHANCEMENTS

* Update Ansible to 2.10 ([GH-136](https://github.com/ystia/forge/issues/136))
* Add docker container property to set the shared memory size ([GH-129](https://github.com/ystia/forge/issues/129))

### BUG FIXES

* Components installing and using the latest pip version fail on systems using python 2.7 ([GH-141](https://github.com/ystia/forge/issues/141))
* Upgrade of ansible to 2.10.0 fails ([GH-138](https://github.com/ystia/forge/issues/138))
* Installation of Consul and Ansible fails on recent Centos for GCP images ([GH-131](https://github.com/ystia/forge/issues/131))

## 2.2.0 (April 17, 2020)

### NEW COMPONENTS

* Failure starting a docker container in attached mode ([GH-119](https://github.com/ystia/forge/issues/119))
* Refactor HAProxy component to take advantage of tosca.nodes.LoadBalancer component ([GH-116](https://github.com/ystia/forge/issues/116))
* SSL Root CA installer component ([GH-103](https://github.com/ystia/forge/issues/103))
* SSL Certificates Generator component ([GH-78](https://github.com/ystia/forge/issues/78))
* Deploy Docker containers in pure TOSCA ([GH-76](https://github.com/ystia/forge/issues/76))
* Docker Registry ([GH-80](https://github.com/ystia/forge/issues/80))
* Expose Hybrid demo material ([GH-93](https://github.com/ystia/forge/issues/93))
* Component for the Yorc plugin provided by Alien4cloud (the Yorc Provider) ([GH-101](https://github.com/ystia/forge/issues/101))
* Docker container component for Alien4Cloud deployment ([GH-111](https://github.com/ystia/forge/issues/111))
* Docker container component for Yorc&Consul deployment ([GH-111](https://github.com/ystia/forge/issues/111))

### NEW TOPOLOGY TEMPLATES

* Template allowing to deploy a simple Yorc+Alien4Cloud stack based on Docker components ([GH-111](https://github.com/ystia/forge/issues/111))

### ENHANCEMENTS

* Add docker container properties workdir and user ([GH-124](https://github.com/ystia/forge/issues/124))
* Update to Kubernetes v1.17 ([GH-117](https://github.com/ystia/forge/issues/117))
* Update Traefik to v2.1 ([GH-115](https://github.com/ystia/forge/issues/115))
* Should support OpenStack Identity API v3 properties ([GH-109](https://github.com/ystia/forge/issues/109))
* Support of locations ([GH-106](https://github.com/ystia/forge/issues/106))
* Update forge components installing/configuring Yorc to support python3 ([GH-85](https://github.com/ystia/forge/issues/85))
* Update Ansible version to 2.7.9 ([GH-83](https://github.com/ystia/forge/issues/83))
* Add a vault in secured full stack topology ([GH-67](https://github.com/ystia/forge/issues/67))
* Provide forge components/topology allowing to install a secured Yorc setup ([GH-37](https://github.com/ystia/forge/issues/37))
* Allow to configure resources prefix for bootstrapped Yorc ([GH-399](https://github.com/ystia/yorc/issues/399))
* Allow to configure Alien4Cloud docker conatainer when deployed within a Yorc stack ([GH-121](https://github.com/ystia/forge/issues/121))

### DEPENDENCIES

* Ystia Forge components require now Alien4Cloud 2.2.0 ([GH-95](https://github.com/ystia/forge/issues/GH-95))
* Ystia Forge components require now Alien4Cloud CSAR public library 2.2.0
* Ystia Forge components require now yorc-types 1.1.0 (GH-82](https://github.com/ystia/forge/pull/82))

### BUG FIXES

* Bad cleanup for components used in yorc bootstrap ([GH-91](https://github.com/ystia/forge/issues/91))
* Empty playbook configure_hostspool_secrets.yml causes error ([GH-88](https://github.com/ystia/forge/issues/88))
* Install a newer version of Anaconda (Python Component) in order to fix dependencies issues ([GH-69](https://github.com/ystia/forge/issues/69))
* Several wrong version referenced for csar-public-lib components ([GH-58](https://github.com/ystia/forge/issues/58))
* Jupyter depends on EPEL repository but do not ensure that it is installed ([GH-70](https://github.com/ystia/forge/issues/70))
* Kubernetes installation fails ([GH-97](https://github.com/ystia/forge/issues/97))
* Fix Kubernetes install if no proxy env vars are defined ([GH-99](https://github.com/ystia/forge/issues/99))
* Bootstrap on OpenStack doesn't allow floating IP provisioning ([GH-516](https://github.com/ystia/yorc/issues/516))
* Yorc Bootstrap doesn't uninstall yorc binary ([GH-605](https://github.com/ystia/yorc/issues/605))

## 2.1.0 (December 20, 2018)

### BUG FIXES

* Kafka download URLs no more valid ([GH-41](https://github.com/ystia/forge/issues/41))
* Types definition issues detected when uploading forge github repository in Alien4Cloud ([GH-42](https://github.com/ystia/forge/issues/42))
* Forge components installing dnsmaq fail on RHEL 7.5  ([GH-44](https://github.com/ystia/forge/issues/44))

### DEPENDENCIES

* Ystia Forge components require now Alien4Cloud 2.1.0
* Ystia Forge components require now Alien4Cloud CSAR public library 2.1.0

## 2.1.0-RC1 (December 17, 2018)

### DEPENDENCIES

* Ystia Forge components require now Alien4Cloud 2.1.0-RC1
* Ystia Forge components require now Alien4Cloud CSAR public library 2.1.0-RC1

### BUG FIXES

* Component NFS Client fails to mount directory exported by NFS server on GCP ([GH-38](https://github.com/ystia/forge/issues/38))

## 2.1.0-M7 (December 07, 2018)

### NEW FEATURES


org/ystia/kibana/linux/bash/scripts/kibana_install.sh
#!/bin/bash
#
# Ystia Forge
# Copyright (C) 2018 Bull S. A. S. - Bull, Rue Jean Jaures, B.P.68, 78340, Les Clayes-sous-Bois, France.
# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
#


source ${utils_scripts}/utils.sh
log begin

# To set variables for the proxy
ensure_home_var_is_set

if isServiceInstalled; then
    log end "Kibana component '${NODE}' already installed"
    exit 0
fi

KBN_ZIP_NAME="kibana-${KBN_VERSION}-linux-x86_64.tar.gz"
KBN_DOWNLOAD_PATH="${REPOSITORY}/${KBN_ZIP_NAME}"
KBN_UNZIP_FOLDER="kibana-${KBN_VERSION}-linux-x86_64"
INSTALL_DIR=${HOME}
KIBANA_HOME_DIR=${INSTALL_DIR}/${KBN_UNZIP_FOLDER}

echo "KIBANA_HOME=${KIBANA_HOME_DIR}" >${YSTIA_DIR}/kibana_env.sh
echo "KIBANA_VERSION=${KBN_VERSION}" >>${YSTIA_DIR}/kibana_env.sh

KIBANA_PLUGIN_DIR=${KIBANA_HOME_DIR}/plugins

if [[ $(majorVersion ${KBN_VERSION}) == 6 ]]
then
    NETWORK_VERSION="v6_2017_12_14"
    HEALTH_METRIC_VERSION="6.2.2"
    SWIMLANE_VERSION="6.2.2"
else
    NETWORK_VERSION="v56_2017_10_27_patched"
fi
#https://github.com/dlumbrer/kbn_network
NETWORK_ZIP_NAME="kbn_network_vis-${NETWORK_VERSION}.zip"
#https://github.com/clamarque/Kibana_health_metric_vis
HEALTH_METRIC_ZIP_NAME="kbn_health_metric_vis-${HEALTH_METRIC_VERSION}.zip"
#https://github.com/prelert/kibana-swimlane-vis
SWIMLANE_ZIP_NAME="kbn_swimlane_vis-${SWIMLANE_VERSION}.zip"


#Install dependencies
sudo yum -y install unzip wget

# Kibana installation
log info "Installing Kibana on ${KIBANA_HOME_DIR}"
( cd ${INSTALL_DIR} && wget ${KBN_DOWNLOAD_PATH} ) || error_exit "ERROR: Failed to install Kibana (download problem) !!!"
if [[ $(majorVersion ${KBN_VERSION}) == 6 ]]
then
    wget ${KBN_DOWNLOAD_PATH}.sha512 -O ${HOME}/${KBN_ZIP_NAME}.sha512
    (cd ${HOME}; sha512sum -c ${KBN_ZIP_NAME}.sha512) || error_exit "ERROR: Checksum validation failed for downloaded Kibana binary"
else
    wget ${KBN_DOWNLOAD_PATH}.sha1 -O ${HOME}/${KBN_ZIP_NAME}.sha1
    (cd ${HOME}; echo "$(cat ${KBN_ZIP_NAME}.sha1) ${KBN_ZIP_NAME}" | sha1sum -c) || error_exit "ERROR: Checksum validation failed for downloaded Kibana binary"
fi
tar -xzf ${INSTALL_DIR}/${KBN_ZIP_NAME} -C  ${INSTALL_DIR} || error_exit "ERROR: Failed to install Kibana (untar problem) !!!"


#
# Kibana plugins installation
#

# args:
# $1 :plugin name
# $2 :plugin zip name
function install_plugin {
    if [[ -e ${plugins}/${2} ]] ; then
        log info ">>>> Install ${1} plugin"
        sudo cp ${plugins}/${2} ${KIBANA_PLUGIN_DIR}
        sudo unzip ${KIBANA_PLUGIN_DIR}/${2} -d ${KIBANA_PLUGIN_DIR}
    else
        log info ">>>> Plugin ${1} missing. Skipping its installation."
    fi
}

# Network plugin installation
install_plugin "Network" ${NETWORK_ZIP_NAME}
if [[ $(majorVersion ${KBN_VERSION}) == 6 ]]
then
    # Health Metric plugin installation
    install_plugin "Health_Metric" ${HEALTH_METRIC_ZIP_NAME}
    # Swimlane plugin installation
    install_plugin "SwimLane" ${SWIMLANE_ZIP_NAME}
fi


# Installation end
setServiceInstalled
log end "Kibana successfully installed"


org/ystia/python/linux/bash/scripts/anaconda_install.sh
#!/usr/bin/env bash
#
# Ystia Forge
# Copyright (C) 2018 Bull S. A. S. - Bull, Rue Jean Jaures, B.P.68, 78340, Les Clayes-sous-Bois, France.
# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
#


function install_pkg_in_mode_off() {
    REPO=$1
    PKG=$2
    cd $HOME
    wget ${REPO}/${PKG}
    conda install -y ${PKG}
    rm -f ${PKG}
    log info ">>>> Conda Package '${PKG}' installed"

}

source ${utils_scripts}/utils.sh
# To set variables for the proxy
ensure_home_var_is_set

log begin

if isServiceInstalled; then
    log end "Python component '${NODE}' already installed"
    exit 0
fi


INSTALL_DIR=$HOME/anaconda2
ANACONDA_INSTALLER_NAME="Anaconda2-2018.12-Linux-x86_64.sh"


# Install bzip2 for anaconda and wget
sudo yum install -y bzip2 wget
log info "bzip2 and wget installed"


if [[ -n "${REPOSITORY}" ]] && [[ "${REPOSITORY}" != "DEFAULT" ]] && [[ "${REPOSITORY}" != "null" ]]; then
    # MODE OFFLINE
    ANACONDA_DOWNLOAD_PATH=${REPOSITORY}/${ANACONDA_INSTALLER_NAME}
else
    # MODE ONLINE
    ANACONDA_DOWNLOAD_PATH="http://repo.continuum.io/archive/${ANACONDA_INSTALLER_NAME}"
fi
# Download and install anaconda
wget "${ANACONDA_DOWNLOAD_PATH}" -P "$HOME" -N
log info "Anaconda Downloaded from ${ANACONDA_DOWNLOAD_PATH}"
bash $HOME/${ANACONDA_INSTALLER_NAME} -b -p ${INSTALL_DIR}
log info "Anaconda Installed"
# Add to path
sudo bash -c "echo 'export PATH="$INSTALL_DIR/bin:$PATH"' >> /etc/bashrc"
log info "PATH Edited"
source /etc/bashrc


if [[ -n "${REPOSITORY}" ]] && [[ "${REPOSITORY}" != "DEFAULT" ]] && [[ "${REPOSITORY}" != "null" ]]; then
    # MODE OFFLINE

    if [[ "${NLP_TWITTER}" == "true" ]]
    then
        for PACKAGE in oauthlib-2.0.6-py_0.tar.bz2 requests-oauthlib-0.8.0-py27_1.tar.bz2 nltk-3.2.5-py27hec5f4de_0.tar.bz2 twython-3.6.0-py27_0.tar.bz2
        do
            install_pkg_in_mode_off ${REPOSITORY} ${PACKAGE}
        done
        log info "nltk and twython installed"
    fi

    if [[ "${DATAVIZ}" == "true" ]]
    then
        sudo yum install -y python-qt4
        for PACKAGE in seaborn-0.8.1-py27_0.tar.bz2 plotly-2.4.1-py27_0.tar.bz2
        do
            install_pkg_in_mode_off ${REPOSITORY} ${PACKAGE}
        done
        log info "seaborn and plotly installed"
    fi

    if [[ "${DATAFORMAT}" == "true" ]]
    then
        for PACKAGE in csvkit-1.0.3-py27_0.tar.bz2 configparser-3.5.0b2-py27_0.tar.bz2
        do
            install_pkg_in_mode_off ${REPOSITORY} ${PACKAGE}
        done
        log info "csvkit and configparser installed"
    fi

    if [[ "${PYBRAIN}" == "true" ]]
    then
        install_pkg_in_mode_off ${REPOSITORY} pybrain-0.3.3-py27_0.tar.bz2
        log info "pybrain installed"
    fi

    if [[ "${ML}" == "true" ]]
    then
        for PACKAGE in pandas-0.23.0-py27h637b7d7_0.tar.bz2 scikit-learn-0.19.1-py27_nomklh6479e79_0.tar.bz2
        do
            install_pkg_in_mode_off ${REPOSITORY} ${PACKAGE}
        done

        log info "pandas and scikit-learn installed"
    fi

else
    # MODE ONLINE

    if [[ "${NLP_TWITTER}" == "true" ]]
    then
        conda install -y nltk
        conda install -y -c conda-forge twython
        log info "nltk and twython installed"
    fi

    if [[ "${DATAVIZ}" == "true" ]]
    then
        conda install -y seaborn
        conda install -y plotly
        # To avoid ImportError: libXext.so.6: cannot open shared object file: No such file or directory
        sudo yum install -y python-qt4
        log info "seaborn and plotly installed"
    fi

    if [[ "${DATAFORMAT}" == "true" ]]
    then
        conda install -y csvkit
        conda install -y configparser
        log info "csvkit and configparser installed"
    fi

    if [[ "${PYBRAIN}" == "true" ]]
    then
        conda install -y -c mq pybrain
        log info "pybrain installed"
    fi

    if [[ "${ML}" == "true" ]]
    then
        conda install -y pandas
        log info "pandas installed"

        conda install -y scikit-learn
        log info "scikit-learn installed"
    fi
fi


setServiceInstalled
log end "Python Installed !"


org/ystia/samples/tcp-echo/ansible/playbooks/files/echo-server.py
from __future__ import print_function
import socket
import sys

# Create a TCP/IP socket
sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

# Bind the socket to the port
server_address = ('', int(sys.argv[1]))
print('starting up on {} port {}'.format(
    server_address[0], server_address[1]), file=sys.stderr)
sock.bind(server_address)

# Listen for incoming connections
sock.listen(1)

while True:
    # Wait for a connection
    print('waiting for a connection', file=sys.stderr)
    connection, client_address = sock.accept()
    try:
        print('connection from', client_address, file=sys.stderr)

        # Receive the data in small chunks and retransmit it
        while True:
            data = connection.recv(16)
            print('received "%s"' % data, file=sys.stderr)
            if data:
                print('sending data back to the client', file=sys.stderr)
                connection.sendall(data)
            else:
                print('no more data from', client_address, file=sys.stderr)
                break

    finally:
        # Clean up the connection
        connection.close()


org/ystia/logstash/linux/bash/relationships/elasticsearch/configure.sh
#!/usr/bin/env bash
#
# Ystia Forge
# Copyright (C) 2018 Bull S. A. S. - Bull, Rue Jean Jaures, B.P.68, 78340, Les Clayes-sous-Bois, France.
# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
#

source ${utils_scripts}/utils.sh
log begin

lock "$(basename $0)"

[ -z ${LOGSTASH_HOME} ] && error_exit "LOGSTASH_HOME not set"

if [[ -e "${YSTIA_DIR}/.${SOURCE_NODE}-ls2esFlag" ]]; then
    log info "Logstash component '${SOURCE_NODE}' already configured with Elasticsearch output"
    unlock "$(basename $0)"
    exit 0
fi


log info "Connecting Logstash to Elasticsearch"
log info "Environment variables : source is ${scripts} - home is ${HOME}"
log info "Elasticsearch cluster name is ${cluster_name}"

# Create conf directory if not already created
mkdir -p ${LOGSTASH_HOME}/conf                                                                                  /
touch "${LOGSTASH_HOME}/conf/3-elasticsearch_logstash_outputs.conf"
# Take into account elk-all-in-one topo. Is consul agent present ?
IS_CONSUL=1
if is_port_open "127.0.0.1" "8500"
then
    host_name="elasticsearch.service.ystia"
elif [[ -n "${ES_PUBLIC_IP}" ]]; then
    host_name="${ES_PUBLIC_IP}"
elif [[ -n "${ES_IP}" ]]; then
    host_name="${ES_IP}"
elif [[ -n "${ES_CAP_IP}" ]]; then
    host_name="${ES_CAP_IP}"
else
    host_name="localhost"
fi
port="9200"
host=$host_name:$port
log info "Elasticsearch host is $host"

if [[ -n "${PROXY}" ]]; then
    proxy_config="\n\t\tproxy => \"${PROXY}\""
fi

#echo -e "output {\n\telasticsearch {\n\t\tcluster => \"${cluster_name}\"\n\t\tprotocol => node\n\t}\n}" >>${LOGSTASH_HOME}/conf/3-elasticsearch_logstash_outputs.conf
echo -e "output {\n\telasticsearch {${proxy_config}\n\t\thosts => [\"$host\"] \n\t}\n}" >>${LOGSTASH_HOME}/conf/3-elasticsearch_logstash_outputs.conf
log info "A4C configure elasticsearch cluster ${cluster_name}"

touch "${YSTIA_DIR}/.${SOURCE_NODE}-ls2esFlag"
log info "Logstash connected to Elasticsearch host ${host}"

unlock "$(basename $0)"

log end

org/ystia/jupyter/linux/bash/scripts/jupyter_create.sh
#!/usr/bin/env bash
#
# Ystia Forge
# Copyright (C) 2018 Bull S. A. S. - Bull, Rue Jean Jaures, B.P.68, 78340, Les Clayes-sous-Bois, France.
# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
#



source ${utils_scripts}/utils.sh
# To set variables for the proxy
ensure_home_var_is_set

log begin

source /etc/bashrc


if isServiceInstalled; then
  echo "Jupyter component already installed"
  exit 0
fi

ENV_JUPYTER_FILE=$HOME/.jupyter/ystia_env4jupyter.sh

sudo yum install -y wget epel-release

jupyter notebook --generate-config
echo "c.NotebookApp.ip = '*'" >> $HOME/.jupyter/jupyter_notebook_config.py
echo "c.NotebookApp.open_browser = False" >> $HOME/.jupyter/jupyter_notebook_config.py
echo "c.NotebookApp.port = 9888" >> $HOME/.jupyter/jupyter_notebook_config.py
# TODO find a way to manage security
echo "c.NotebookApp.token = ''" >> $HOME/.jupyter/jupyter_notebook_config.py
echo "c.NotebookApp.password = ''" >> $HOME/.jupyter/jupyter_notebook_config.py


# Configure default directory for notebooks storage
mkdir $HOME/MyNotebooks
echo "c.NotebookApp.notebook_dir='"$HOME/MyNotebooks"'" >> $HOME/.jupyter/jupyter_notebook_config.py


if [[ "${IRKERNEL}" == "true" ]]
then
  #base
  sudo yum install -y libzmq3-dev python-zmq
  sudo yum install -y zeromq-devel
  #gitr2 error donc pas devtools
  sudo yum install -y openssl-devel
  sudo yum install -y czmq-devel
  #Unable to find the LibSSH2 library on this mais pas d'erreur
  sudo yum install -y libssh2-devel
  #curl
  sudo yum install -y libcurl-devel

  log info "R installation"
  sudo yum install -y R
  log info "R is installed"
  #install dans  nouvelle library
  sudo bash -c "echo 'export LD_LIBRARY_PATH="/lib/:/lib64/:/usr/lib64/:/usr/lib64/R/lib:$HOME/anaconda2/lib"' >> /etc/bashrc"
  #Fix issue BRBDCF-466
  mkdir -p $HOME/R/x86_64-redhat-linux-gnu-library/3.3
  echo "R_LIBS=$HOME/R/x86_64-redhat-linux-gnu-library/3.3" >>${ENV_JUPYTER_FILE}

  #get the export before Rscript
  source /etc/bashrc
  log info "add new library"
  Rscript ${data_file}/packJup.R

  log info "irkernel installed"
fi

if [[ "${PY35KERNEL}" == "true" ]]
then
  echo y | conda create -n py35 python=3.5 anaconda
  log info "env for python35 created"
  source activate py35
  #otherwise permission denied
  sudo bash -c 'source /etc/bashrc; ipython kernel install'
  source deactivate
  # To see py35 in the 'New' menu of Notebook
  # Cf https://stackoverflow.com/questions/30492623/using-both-python-2-x-and-python-3-x-in-ipython-notebook
  cp -r $HOME/anaconda2/envs/py35/share/jupyter/kernels/python3 $HOME/anaconda2/share/jupyter/kernels
  log info "py35 kernel installed"
fi


# TODO Will be taken into account when MapR/Hortonworks will be migrated...
if [[ "${SPARK_KERNEL}" == "true" ]]
then

  #MODE OFFLINE
  if [[ -n "${REPOSITORY}" ]] && [[ "${REPOSITORY}" != "DEFAULT" ]] && [[ "${REPOSITORY}" != "null" ]]; then
    #download and install anaconda

    cd $HOME
    wget "${REPOSITORY}/spark-1.5.2-bin-hadoop2.6.tgz" -P "$HOME" -N
    tar zxvf spark-1.5.2-bin-hadoop2.6.tgz
    log info "spark installed"

    wget "${REPOSITORY}/spark-kernel-0.1.5-SNAPSHOT.tar.gz" -P "$HOME" -N
    tar zxvf spark-kernel-0.1.5-SNAPSHOT.tar.gz
    log info "spark-kernel downloaded"

    sudo yum install -y java
    log info "java installed"

    mkdir -p ~/.ipython/kernels/spark
    cp -r spark-kernel ~/.ipython/kernels/spark/
    cd ~/.ipython/kernels/spark/

  else
    #MODE ONLINE
    cd $HOME
    wget "http://d3kbcqa49mib13.cloudfront.net/spark-1.5.2-bin-hadoop2.6.tgz" -P "$HOME" -N
    tar zxvf spark-1.5.2-bin-hadoop2.6.tgz
    log info "spark installed"

    git clone https://github.com/ibm-et/spark-kernel.git

    log info "spark-kernel downloaded"

    cd spark-kernel
    export APACHE_SPARK_VERSION=1.5.2

    curl https://bintray.com/sbt/rpm/rpm | sudo tee /etc/yum.repos.d/bintray-sbt-rpm.repo
    sudo yum install -y sbt
    log info "sbt installed"

    make dist
    log info "make dist"


    mkdir -p ~/.ipython/kernels/spark
    cp -r dist/spark-kernel ~/.ipython/kernels/spark/
    cd ~/.ipython/kernels/spark/

  fi
  #delete the last 3 lines
  head -n -3 ./spark-kernel/bin/spark-kernel > tmp.txt; rm -f ./spark-kernel/bin/spark-kernel; mv tmp.txt ./spark-kernel/bin/spark-kernel

  echo 'exec "$SPARK_HOME"/bin/spark-submit \
  ${SPARK_OPTS} \
  --driver-class-path $PROG_HOME/lib/${KERNEL_ASSEMBLY} \
  --class com.ibm.spark.SparkKernel $PROG_HOME/lib/${KERNEL_ASSEMBLY} "$@"
  ' >> ./spark-kernel/bin/spark-kernel

  echo "{
    \"display_name\": \"Spark\",
    \"language_info\": { \"name\": \"scala\" },
    \"argv\": [
    \"$HOME/.ipython/kernels/spark/spark-kernel/bin/spark-kernel\",
    \"--master\",
    \"yarn-client\",
    \"--profile\",
    \"{connection_file}\"
    ],
    \"codemirror_mode\": \"scala\",
    \"env\": {
      \"SPARK_OPTS\": \"--driver-java-options=-Xms1024M --driver-java-options=-Xmx4096M --driver-java-options=-Dlog4j.logLevel=trace\",
      \"MAX_INTERPRETER_THREADS\": \"16\",
      \"SPARK_CONFIGURATION\": \"spark.cores.max=4\",
      \"CAPTURE_STANDARD_OUT\": \"true\",
      \"CAPTURE_STANDARD_ERR\": \"true\",
      \"SEND_EMPTY_OUTPUT\": \"false\",
      \"SPARK_HOME\": \"$HOME/spark-1.5.2-bin-hadoop2.6\",
      \"PYTHONPATH\": \"$HOME/spark-1.5.2-bin-hadoop2.6/python:$HOME/spark-1.5.2-bin-hadoop2.6/python/lib/py4j-0.9-src.zip\"
    }
  }" > ~/.ipython/kernels/spark/kernel.json

  # otherwise permision denied with jupyter notebook
  sudo chmod 751 -R $HOME/.ipython

fi


# TODO Will be taken into account when H2O will be migrated...
if [[ "${H2O}" == "true" ]]
then
  conda install -y requests
  conda install -y tabulate
  conda install -y scikit-learn
  conda install -y colorama
  conda install -y future
  log info "env for h2o created"
  conda install -y h2o h2o-py
  log info "h2o installed"
fi

# Take into account proxy
[[ -r /etc/profile.d/proxy.sh ]] && (cat /etc/profile.d/proxy.sh | sed -e "s|export||" >>${ENV_JUPYTER_FILE})
# Setup systemd service
sudo cp ${scripts}/systemd/jupyter.service /etc/systemd/system/jupyter.service
sudo sed -i -e "s/{{USER}}/${USER}/g" -e "s@{{CMD_JUPYTER}}@$(which jupyter)@g" /etc/systemd/system/jupyter.service
if [[ -r ${ENV_JUPYTER_FILE} ]]
then
  sudo sed -i -e "s@{{ENVIRONMENTFILE}}@EnvironmentFile=${ENV_JUPYTER_FILE}@g" /etc/systemd/system/jupyter.service
else
  sudo sed -i -e "s@{{ENVIRONMENTFILE}}@@g" /etc/systemd/system/jupyter.service
fi
sudo systemctl daemon-reload
#sudo systemctl enable jupyter.service


setServiceInstalled

log end "Jupyter installed !"
